### PR TITLE
fix(LINGUIST-246): Added logs and global exception handler, removed try catches from the controller and updated try catch blocks in services.

### PR DESCRIPTION
**Pull request by elif-krvan**
Since GE250 users will test our system, we need event logs so that we can understand if something goes wrong. Therefore, please add information and error logs to the services you developed previously. Also, from now on, please remove try-catch blocks from the controller classes and add appropriate throw declarations to the service layer.

### CODE REVIEW

File: src/main/java/app/linguistai/bmvp/service/AccountService.java [hunk 1]
```diff
@@ -69,13 +74,18 @@ public RLoginUser login(QUserLogin user) throws Exception {
             }
             catch (Exception e1) {
                 // Intentionally not thrown to not cause login exception for users without UserStreak
-                System.out.println(ExceptionLogger.log(e1));
+                System.out.println(ExceptionLogger.log(e1)); // TODO what is this
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/AccountService.java:77]: please remove TODO and please add Jira issue for me

File: src/main/java/app/linguistai/bmvp/service/AccountService.java [hunk 2]
```diff
@@ -85,29 +95,24 @@ public RRefreshToken refreshToken(String auth) throws Exception {
...
     }
 
     public boolean changePassword(String email, QChangePassword passwords) throws Exception {
         try {
-            User dbUser = accountRepository.findUserByEmail(email).orElse(null);
-
-            if (dbUser == null) {
-                throw new Exception("User is not found");
-            }
+            User dbUser = accountRepository.findUserByEmail(email).orElseThrow(() -> new NotFoundException("User", true));
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/AccountService.java:108]: `"User"` -> `user.getClass().getSimpleName()`

File: src/main/java/app/linguistai/bmvp/service/AccountService.java [hunk 3]
```diff
@@ -158,7 +180,7 @@ public List<User> getUsers() {
         }
     }
 
-    private String encodePassword(String plainPassword) {
+    private String encodePassword(String plainPassword) { // TODO no need for this method
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/AccountService.java:183]: Please remove TODO and add `@Deprecated` to the method, with a comment saying that we don't need this method. We will remove deprecated code in a future sprint

File: src/main/java/app/linguistai/bmvp/service/AccountService.java [hunk 4]
```diff
@@ -168,11 +190,7 @@ private String encodePassword(String plainPassword) {
 
     public ResetToken generateEmailToken(String email) throws Exception {
         try {
-            User user = accountRepository.findUserByEmail(email).orElse(null);
-
-            if (user == null) {
-                throw new NotFoundException("User with email [" + email + "] not found");
-            }
+            User user = accountRepository.findUserByEmail(email).orElseThrow(() -> new NotFoundException("User", true));
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/AccountService.java:193]: "User" rename as mentioned previously

File: src/main/java/app/linguistai/bmvp/service/AccountService.java [hunk 5]
```diff
@@ -182,39 +200,51 @@ public ResetToken generateEmailToken(String email) throws Exception {
...
             if (invalidate) {
                 resetToken.setUsed(true);
                 resetTokenRepository.save(resetToken);
             }
+
+            log.info("Reset code is validated for user {}", user.getId());
+
             return true;
+        } catch (NotFoundException e) {
+            if (e.getObject().equals("User")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/AccountService.java:235]: defining "User" as a const somewhere may be beneficial, as it looks like it is being used many times

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 6]
```diff
@@ -25,84 +31,100 @@ public class FriendshipService {
...
     public Friendship sendFriendRequest(String user1Email, UUID user2Id) throws Exception {
         try {
-            User dbUser1 = accountRepository.findUserByEmail(user1Email).orElseThrow(() -> new Exception("User is not found"));
-            User dbUser2 = accountRepository.findUserById(user2Id).orElseThrow(() -> new Exception("Requested user is not found"));
+            User dbUser1 = accountRepository.findUserByEmail(user1Email).orElseThrow(() -> new NotFoundException("User", true));
+            User dbUser2 = accountRepository.findUserById(user2Id).orElseThrow(() -> new NotFoundException("Requested user", true));
 
             if (dbUser1.getId().equals(dbUser2.getId())) {
-                throw new Exception("User cannot send friend request to themselves");
+                throw new FriendException("You cannot send friend request to yourself");
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:38]: "cannot send friend request" -> "cannot send `a` friend request"

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 7]
```diff
@@ -25,84 +31,100 @@ public class FriendshipService {
...
 
             // check if friend request exists
             Friendship friendship = friendshipRepository.findByUserPair(dbUser1.getId(), user2Id).orElse(null);
 
             if (friendship != null) {
-                throw new Exception("Friendship or request already exist");
+                if (friendship.getStatus() == FriendshipStatus.PENDING) {
+                    throw new AlreadyFoundException("Friendship request", true);
+                } else {
+                    throw new AlreadyFoundException("Friendship", true);
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:48]: Similar to "User", you can define "Friendship" as a constant

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 8]
```diff
@@ -25,84 +31,100 @@ public class FriendshipService {
...
+                throw new FriendException("You cannot send friend request to yourself");
             }
 
             // check if friend request exists
             Friendship friendship = friendshipRepository.findByUserPair(dbUser1.getId(), user2Id).orElse(null);
 
             if (friendship != null) {
-                throw new Exception("Friendship or request already exist");
+                if (friendship.getStatus() == FriendshipStatus.PENDING) {
+                    throw new AlreadyFoundException("Friendship request", true);
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:46]: Similar to "User", you can define "Friendship request" as a constant

Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:46]: Here, the code checks `friendship.getStatus()` and then based on the result, throws different exceptions to differentiate between the cases. This introduces nesting and also a little bit of repetition.

Utilizing `switch expressions` introduced in Java 14 will significantly reduce nesting and improve the readability and maintainability of the code. Example:

```
switch (number) {
	case 1 -> callOne();
	case 2 -> callTwo();
	default -> callMany();
}
```

See [https://nipafx.dev/java-switch/](https://nipafx.dev/java-switch/) for details.

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 9]
```diff
@@ -25,84 +31,100 @@ public class FriendshipService {
...
             // save friendship to the db with pending status
             friendship = friendshipRepository.save(new Friendship(dbUser1, dbUser2, now, FriendshipStatus.PENDING));
 
+            log.info("Friend request is sent from {} to {}", dbUser1.getId(), dbUser2.getId());
+
             return friendship;
-        } catch (Exception e) {
-            System.out.println(e.getLocalizedMessage());
+        } catch (NotFoundException e) {
+            if (e.getObject().equals("User")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:61]: "User" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 10]
```diff
@@ -25,84 +31,100 @@ public class FriendshipService {
...
+                log.error("User is not found for email {}", user1Email);
+            } else {
+                log.error("Requested user with id {} not found.", user2Id);
+            }
+            throw e;
+        } catch (FriendException e) {
+            log.error("Friend request failed since user with email {} tried to send request to themselves", user1Email);
+            throw e;
+        } catch (AlreadyFoundException e) {
+            if (e.getObject().equals("Friendship")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:71]: "Friendship" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 11]
```diff
@@ -25,84 +31,100 @@ public class FriendshipService {
...
-            } else if (dbUser2 == null) {
-                throw new Exception("Requested user is not found");
-            }
-
-            Friendship friendship = friendshipRepository.findByUser1IdAndUser2IdAndStatus(dbUser1.getId(), dbUser2.getId(), FriendshipStatus.PENDING).orElse(null);
-
-            if (friendship == null) {
-                throw new Exception("Friendship is not found");
-            }
+            Friendship friendship = friendshipRepository.findByUser1IdAndUser2IdAndStatus(dbUser1.getId(), dbUser2.getId(), FriendshipStatus.PENDING).orElseThrow(() -> new NotFoundException("Friendship", true));
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:127]: "Friendship" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 12]
```diff
@@ -112,10 +134,22 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
             // save friendship to the db with pending status
             friendship = friendshipRepository.save(friendship);
 
+            log.info("User {} accepted friend request of {}.", dbUser2.getId(), dbUser1.getId());
+
             return friendship;
-        } catch (Exception e) {
-            System.out.println(e.getMessage());
+        } catch (NotFoundException e) {
+            if (e.getObject().equals("User")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:141]: "User" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 13]
```diff
@@ -112,10 +134,22 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
+                log.error("Accept friend request failed since user does not exists for id {}", user1Id);
+            } else if (e.getObject().equals("Requested user")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:143]: "Requested user" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 14]
```diff
@@ -112,10 +134,22 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
+                log.error("Accept friend request failed since requested user does not exists for email {}", user2Email);
+            } else if (e.getObject().equals("Friendship")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:145]: "Friendship" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 15]
```diff
@@ -124,55 +158,63 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
 
             friendshipRepository.delete(friendship);
 
+            log.info("User {} rejected friend request of {}.", dbUser2.getId(), dbUser1.getId());
+
             return friendship;
-        } catch (Exception e) {
-            System.out.println(e.getLocalizedMessage());
+        } catch (NotFoundException e) {
+            if (e.getObject().equals("User")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:174]: "User" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 16]
```diff
@@ -124,55 +158,63 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
+                log.error("Reject friend request failed since user does not exists for id {}", user1Id);
+            } else if (e.getObject().equals("Requested user")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:176]: "Requested user" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 17]
```diff
@@ -124,55 +158,63 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
+                log.error("Reject friend request failed since requested user does not exists for email {}", user2Email);
+            } else if (e.getObject().equals("Friendship")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:178]: "Friendship" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 18]
```diff
@@ -124,55 +158,63 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
 
             friendshipRepository.delete(friendship);
 
+            log.info("User {} removed their friend {}.", dbUser1.getId(), dbUser2.getId());
+
             return friendship;
-        } catch (Exception e) {
-            System.out.println(e.getLocalizedMessage());
+        } catch (NotFoundException e) {
+            if (e.getObject().equals("User")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:206]: "User" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 19]
```diff
@@ -124,55 +158,63 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
+                log.error("Remove friend failed since user does not exists for id {}", user1Id);
+            } else if (e.getObject().equals("Requested user")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:208]: "Requested user" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 20]
```diff
@@ -124,55 +158,63 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
+                log.error("Remove friend failed since requested user does not exists for email {}", user2Email);
+            } else if (e.getObject().equals("Friendship")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:210]: "Friendship" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/UserStreakService.java [hunk 21]
```diff
@@ -66,9 +72,17 @@ private RUserStreak checkUserStreakForUpdate(RUserStreak rUserStreak) throws Exc
                 // Streak must be reset (edge case, if last login is after current time) Streak must be reset
                 default -> resetUserStreak(userStreak);
             };
-        }
-        catch (Exception e) {
+        } catch (NotFoundException e) {
+            if (e.getObject().equals("StreakTime")) {
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/UserStreakService.java:76]: "StreakTime" should be const

File: src/main/java/app/linguistai/bmvp/service/gamification/UserStreakService.java [hunk 22]
```diff
@@ -42,15 +46,17 @@ private RUserStreak checkUserStreakForUpdate(RUserStreak rUserStreak) throws Exc
             Date streakTime = DateUtils.convertSqlDateToUtilDate(rUserStreak.getLastLogin());
 
             if (streakTime == null) {
-                throw new Exception("StreakTime is null");
+                throw new NotFoundException("StreakTime", true);
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/UserStreakService.java:49]: "StreakTime" should be const

Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/UserStreakService.java:49]: Moreover, this should not be a `NotFoundException`, as for this case the **UserStreak is there**. The problem as that a `NotNull` field is actually `Null`.

**Review by can-aslan — CHANGES_REQUESTED** [review]
Looks very good overall 🚀, just left some minor comments. The count may seem high but most requests are actually duplicates.

File: src/main/java/app/linguistai/bmvp/controller/AccountController.java [hunk 23]
```diff
@@ -42,96 +41,58 @@ public class AccountController {
...
     }
 
     @GetMapping("hello")
     public ResponseEntity<Object> testUnsecuredEndpoint() {
-        try {
-            return Response.create("Ok", HttpStatus.OK);
-        } catch (Exception e) {
-            return Response.create(ExceptionLogger.log(e), HttpStatus.OK);  
-        }        
+        return Response.create("Ok", HttpStatus.OK);
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/controller/AccountController.java:82]: Best endpoint

Comment (can-aslan) [reply src/main/java/app/linguistai/bmvp/controller/AccountController.java:82]: 🥇

File: src/main/java/app/linguistai/bmvp/service/AccountService.java [hunk 24]
```diff
@@ -192,39 +210,51 @@ public ResetToken generateEmailToken(String email) throws Exception {
...
 
-            ResetToken resetToken = resetTokenRepository.findByUserAndResetCode(user, resetCode).orElse(null);
-            if (resetToken == null) {
-                throw new NotFoundException("Reset token for user with email [" + user.getEmail() + "] with code [" + resetCode + "] not found.");
-            }
+            ResetToken resetToken = resetTokenRepository.findByUserAndResetCode(user, resetCode).orElseThrow(() -> new NotFoundException("Reset token", true));
 
             if (!isResetTokenValid(resetToken)) {
-                return false;
+                throw new TokenException("Invalid password reset token");
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/AccountService.java:233]: It looks like there are other functions depending on this method returning a boolean value, including `false` (i.e. `AccountController.saveResetPassword`, `AccountController.validateResetPassword`). Can you please check that no functionality is broken before merging? Thanks

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java (hunk 8 above)
Comment (can-aslan) [reply src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:48]: @elif-krvan looks like this remained like hard-coded string, can you check?

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java (hunk 9 above)
Comment (can-aslan) [reply src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:63]: @elif-krvan looks like this remained like hard-coded string, can you check?

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 25]
```diff
@@ -124,55 +160,63 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
-            Friendship friendship = friendshipRepository.findByUserPairAndStatus(dbUser1.getId(), dbUser2.getId(), FriendshipStatus.PENDING).orElse(null);
+            User dbUser1 = accountRepository.findUserById(user1Id).orElseThrow(() -> new NotFoundException("User", true));
+            User dbUser2 = accountRepository.findUserByEmail(user2Email).orElseThrow(() -> new NotFoundException(REQ_USER_STR, true));
 
-            if (friendship == null) {
-                throw new Exception("Friendship is not found");
-            }
+            Friendship friendship = friendshipRepository
+                                    .findByUserPairAndStatus(dbUser1.getId(), dbUser2.getId(), FriendshipStatus.PENDING)
+                                    .orElseThrow(() -> new NotFoundException("Friendship", true));
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:168]: hard-coded string

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 26]
```diff
@@ -124,55 +160,63 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
-            Friendship friendship = friendshipRepository.findByUserPairAndStatus(dbUser1.getId(), dbUser2.getId(), FriendshipStatus.ACCEPTED).orElse(null);
+            User dbUser1 = accountRepository.findUserById(user1Id).orElseThrow(() -> new NotFoundException("User", true));
+            User dbUser2 = accountRepository.findUserByEmail(user2Email).orElseThrow(() -> new NotFoundException(REQ_USER_STR, true));
 
-            if (friendship == null) {
-                throw new Exception("Friendship is not found");
-            }
+            Friendship friendship = friendshipRepository
+                                    .findByUserPairAndStatus(dbUser1.getId(), dbUser2.getId(), FriendshipStatus.ACCEPTED)
+                                    .orElseThrow(() -> new NotFoundException("Friendship", true));
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:200]: hard-coded string

File: src/main/java/app/linguistai/bmvp/service/gamification/UserStreakService.java [hunk 27]
```diff
@@ -129,7 +146,7 @@ public RUserStreak getUserStreakByToken(String email) throws Exception {
             throw e1;
         }
         catch (Exception e2) {
-            System.out.println("ERROR: Could not fetch UserStreak.");
+            System.out.println("ERROR: Could not fetch UserStreak."); // TODO what is this
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/UserStreakService.java:149]: please remove todo (discussed already, will be addressed with [LINGUIST-350](https://linguistai.atlassian.net/browse/LINGUIST-350))

File: src/main/java/app/linguistai/bmvp/service/profile/ProfileService.java [hunk 28]
```diff
@@ -47,30 +52,25 @@ private WebClient getWebClient() {
     // this method should be called only when a new user is created
     public boolean createEmptyProfile(UUID userId) throws Exception {
         try {
-            UserProfile dbProfile = profileRepository.findById(userId).orElse(null);
-
-            if (dbProfile == null) {
-                throw new Exception("User profile already exists!");
-            }
+            profileRepository.findById(userId).orElseThrow(() -> new AlreadyFoundException("User profile", true));
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/profile/ProfileService.java:55]: hard-coded string

**Review by can-aslan — CHANGES_REQUESTED** [review]
LGTM, left some minor comments

File: src/main/java/app/linguistai/bmvp/service/AccountService.java (hunk 24 above)
Comment (elif-krvan) [reply src/main/java/app/linguistai/bmvp/service/AccountService.java:233]: offffffffffffffffffffffffffffffffffffffffffffffffffff

File: src/main/java/app/linguistai/bmvp/controller/AccountController.java (hunk 23 above)
Comment (elif-krvan) [reply src/main/java/app/linguistai/bmvp/controller/AccountController.java:82]: sanane

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 29]
```diff
@@ -124,55 +161,63 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
-            User dbUser2 = accountRepository.findUserByEmail(user2Email).orElse(null);
-
-            if (dbUser1 == null) {
-                throw new Exception("User is not found");
-            } else if (dbUser2 == null) {
-                throw new Exception("Requested user is not found");
-            }
-
-            Friendship friendship = friendshipRepository.findByUserPairAndStatus(dbUser1.getId(), dbUser2.getId(), FriendshipStatus.PENDING).orElse(null);
+            User dbUser1 = accountRepository.findUserById(user1Id).orElseThrow(() -> new NotFoundException("User", true));
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:164]: hard coded string "User"

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 30]
```diff
@@ -124,55 +161,63 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
-            User dbUser2 = accountRepository.findUserByEmail(user2Email).orElse(null);
-
-            if (dbUser1 == null) {
-                throw new Exception("User is not found");
-            } else if (dbUser2 == null) {
-                throw new Exception("Requested user is not found");
-            }
-
-            Friendship friendship = friendshipRepository.findByUserPairAndStatus(dbUser1.getId(), dbUser2.getId(), FriendshipStatus.ACCEPTED).orElse(null);
+            User dbUser1 = accountRepository.findUserById(user1Id).orElseThrow(() -> new NotFoundException("User", true));
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:196]: hard coded string "User"

File: src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java [hunk 31]
```diff
@@ -124,55 +161,63 @@ public Friendship acceptRequest(String user2Email, UUID user1Id) throws Exceptio
...
+            User dbUser2 = accountRepository.findUserByEmail(user2Email).orElseThrow(() -> new NotFoundException(REQ_USER_STR, true));
 
-            if (friendship == null) {
-                throw new Exception("Friendship is not found");
-            }
+            Friendship friendship = friendshipRepository
+                                    .findByUserPairAndStatus(dbUser1.getId(), dbUser2.getId(), FriendshipStatus.ACCEPTED)
+                                    .orElseThrow(() -> new NotFoundException("Friendship", true));
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/gamification/FriendshipService.java:201]: hard coded string "Friendship"

File: src/main/java/app/linguistai/bmvp/service/profile/ProfileService.java [hunk 32]
```diff
@@ -126,22 +126,22 @@ public RUserProfile updateUserProfile(String email, QUserProfile profile) throws
...
 
     public RUserProfile getUserProfile(String email) throws Exception {
         try {
             // check if user exists
-            User dbUser = accountRepository.findUserByEmail(email).orElse(null);
-
-            if (dbUser == null) {
-                throw new Exception("User does not exist!");
-            }
+            User dbUser = accountRepository.findUserByEmail(email).orElseThrow(() -> new NotFoundException("User", true));
```
Comment (can-aslan) [inline src/main/java/app/linguistai/bmvp/service/profile/ProfileService.java:144]: hard coded string "User"

**Review by can-aslan — CHANGES_REQUESTED** [review]

**Review by can-aslan — APPROVED** [review]
LGTM, baya işe yarayacak bir PR, eline sağlıkk 🚀 🚀 🚀 🚀 🚀

**Pull request merged**